### PR TITLE
Ma va item alert list

### LINKED ISSUE #6: 6. Como usuario, quiero ser alertado en el momento de que estoy ingresando un item que es igual a algo ya existente en la lista de compras, así puedo evitar items duplicados
AC:

* Muestra un mensaje de error si el usuario trata de enviar un item que tiene un nombre exactamente igual a otro ya existente
* Muestra un mensaje de error si el usuario trata de ingresar un nuevo item que tiene el mismo nombre de un item ya existente, después de normalizar mayúsculas y remover signos de puntuación
* El input original del usuario es guardado en la base de datos

### PR DESCRIPTION
**Pull request by Lukaina**
## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|  ✓ | :sparkles: New feature     |

## Description

As a user, I want to be alerted when I’m entering an item that is the same as something already on my list so I can avoid duplicate items.

### To verify its use:
1. Add a product in the input (it will be saved in the database)
2. If you try to add the same product, the application will give you an error message because the product has already been added. 

## Related Issue

Fix #6 

## Acceptance Criteria

- Show an error message if the user tries to submit a new item that has the exact same name as an existing item
- Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
- The user’s original input should be what gets saved in the database

## Captures
capture

### CODE REVIEW

**Comment by alxmcr** [comment]
Good job @lukaina and @macosta08! 

Well, about my revision:
- Created a new collection "`alejandro`" -> OK
- Add Items -> OK

Also, I tried to add items with the same name, even though with names in uppercase or lowercase in all these cases **_I got an error validation. Great!_**, I read your code, I could identify many logic, much word processing, and a big work! **Congratulations!**

But, I tried with values maybe edges cases, because I could add items with the same names with punctuation. Except when punctuation between other letters (For example: `orange.` and `oran.ge` / `Ap.ple` and `apple.`)

image


I don't know, **_is it possible with actual code?_** or maybe I can be confused with the second acceptance criteria. Sorry about it. Please, I'll appreciate some clarification.

**That's all! thanks for your effort and time!.**

File: src/components/FormProduct/index.jsx
Comment (alxmcr) [inline src/components/FormProduct/index.jsx]: When I opened the browser console then it showed me a **_message_** `"String prototype is read only, properties should not be added  no-extend-native".` It's a **warning**, maybe you can create an external function for your "normalizeItem" method if it's possible.

image

**Comment by Lukaina** [comment]
@alxmcr, Thank you very much for your comments and suggestions! Everything is of great help for the application to improve. Good thing you read the code, tested it and it worked. Well, reviewing the issue and in one part it says: "remove punctuation marks", I will ask the question in Slack to our mentors to give clarity, as it has also generated me confusion and given the case, the change is made (Thanks for testing that).

Thanks to you for taking the time to review and for all the support.

File: src/components/FormProduct/index.jsx
Comment (macosta08) [reply src/components/FormProduct/index.jsx]: - Hi Ale, it seems like an excellent test case and it would also meet the acceptance criteria. I made the change, thank you very much. 
- I exported the method nomalizaItem to another file. The warning appears the same. It is a question that I have pending to ask. Why does the warning appear if it is a JavaScript functionality?

Comment (alxmcr) [reply src/components/FormProduct/index.jsx]: Hi @macosta08 ! Well, I tried with `grapes.` and `Gr.apes` and now I got a validation error. Great! thanks for fix it!

Well, about your function with `String.prototype.normalizeItem` Is it necessary to use this structure? is not possible functions like in `firestore.js` ? I believe the warning is because your only "read" but not modified. On another hand I tried to find some answer about or you can ask to mentors, please!

**Review by CrisRuedaP — APPROVED** [review]
I've tested the feature and everything it's working fine. Nice job!

File: src/components/FormProduct/index.jsx
Comment (Lukaina) [reply src/components/FormProduct/index.jsx]: Hi @alxmcr, the recommendations you make are great, they help us a lot to learn and improve, as my colleague explained, it would meet the acceptance criteria. By the way @macosta08, thank you very much for making the clarification and generating the corresponding changes. I see that normalizeItem has already been extracted to the firestore.js file, thank you both very much!

Comment (Lukaina) [reply src/components/FormProduct/index.jsx]: Hi @alxmcr, the recommendations you make are great, they help us a lot to learn and improve, as my colleague explained, it would meet the acceptance criteria. You can verify that the warning message is no longer displayed. By the way @macosta08, thank you very much for making the clarification and generating the corresponding changes. I see that normalizeItem has already been extracted to the firestore.js file, thank you both very much!

**Review by alxmcr — APPROVED** [review]
Thanks, @macosta08  and @Lukaina for your corrections and support! 

I reviewed:
- Avoid duplicates -> OK
- Warning, function -> OK

Approved! ✔️

File: src/components/Utils/normalizeItem.js
Comment (alejandronanez) [inline src/components/Utils/normalizeItem.js]: This is a big no-no as it could be really dangerous to modify an object's prototype.

Let's create this as a regular function instead. It could live in a `/helpers` file or something like that.

**Review by alejandronanez — CHANGES_REQUESTED** [review]
Hey y'all - on my phone right now, sorry for the brevity on my review. But I wanted to point out something that needs to be changed before anything else.

Ty!

File: src/components/Utils/normalizeItem.js
Comment (Lukaina) [reply src/components/Utils/normalizeItem.js]: @alejandronanez Thank you very much for the suggestion! I didn't know that you had to be careful with the prototype, it's still not very clear to me why. I continue to learn every day and it's great. The corresponding changes have already been made.

Comment (macosta08) [reply src/components/Utils/normalizeItem.js]: Ok, we already change it to a regular function. @alejandronanez it would be really nice if you could explain us or give us more details why is a big no-no :). Tnahks in advance.

Comment (indiragp) [reply src/components/Utils/normalizeItem.js]: This resource has a pretty good answer: https://flaviocopes.com/javascript-why-not-modify-object-prototype/ 
Also, please remove this file :)

File: src/components/FormProduct/index.jsx
Comment (indiragp) [inline src/components/FormProduct/index.jsx]: I would suggest sending the item as is to the productExists function, and inside that function do the normalizeItem for the input. 
This will result in cleaner code, since you are only doing normalization in a single place, and calling and importing the normalizeItem function in one single file.

File: src/components/Utils/normalizeItem.js
Comment (Lukaina) [reply src/components/Utils/normalizeItem.js]: @indiragp, Thank you very much for the resource! The file has already been removed, it has been renamed as helpers.js taking into account the suggestion of @alejandronanez, if you want you can see the latest changes.

File: src/components/FormProduct/index.jsx
Comment (macosta08) [reply src/components/FormProduct/index.jsx]: Done...

File: src/components/Utils/normalizeItem.js
Comment (alejandronanez) [reply src/components/Utils/normalizeItem.js]: Sorry y'all for leaving you hanging a couple of days ago 😅, I was on a hurry and from my 📱.

The article @indiragp shared is great, bookmark it for future reference :)

This operation is a big no-no because once you modify an Object's Prototype you _hide_ information and could easily override an existing method, for instance:

```js
// Array prototype has a `map` method

[1,2,3].map(n => n +1) // [2,3,4]

// But, if you decide to implement your own map like this
Array.prototype.map = () => 'hello';
// and then you do
[1,2,3].map(n => n + 1)
// You'll get "hello" back!
```

Happy to talk more about this if you still have questions! 😄 @Lukaina @macosta08

File: src/components/FormProduct/index.jsx
Comment (alejandronanez) [inline src/components/FormProduct/index.jsx:24]: Nice, thanks from changing this from `let => const`

File: src/components/FormProduct/index.jsx
Comment (alejandronanez) [inline src/components/FormProduct/index.jsx]: This comment is a bit redundant, and redundant comments should be avoided if possible.

What I think that you could do it to rename `productExists` to `isProductDuplicated` and then remove the comment above it. If you do that, then the function name will imply that we're checking for duplicates.

File: src/components/Utils/helpers.js
Comment (alejandronanez) [inline src/components/Utils/helpers.js]: Here's you're passing `String` as an argument, and this works, but, `String` is already an [Object declared by Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String) 

Let's change this to something different.

I'd suggest you changing this to `item` or `itemName`.

```js
// Something like this should work.
export function normalizaItem(item) { ... }
```

File: src/components/FormProduct/index.jsx
Comment (alejandronanez) [inline src/components/FormProduct/index.jsx]: I don't really understand this comment 😅.

Maybe should it be before the `setError` call? Maybe it should be more explicit? No biggie though, but, still worth checking out.

**Review by alejandronanez — APPROVED** [review]
**☝️ I didn't QA this feature, so do not merge it before getting an approval from your mentors. ☝️**

I liked the code I saw here, nice work y'all. Please, consider doing the changes I suggested in my review. 

Thank you! 👍

File: src/components/Utils/normalizeItem.js
Comment (macosta08) [reply src/components/Utils/normalizeItem.js]: Thanks a lot @alejandronanez @indiragp 👍

**Pull request merged**
